composer/loggers/wandb_logger.py

```diff
@@ -89,30 +90,45 @@ def state_dict(self) -> Dict[str, Any]:
         import wandb
         del state  # unused
-        if "name" not in self._init_params:
-            # Use the logger run name if the name is not set.
+
+        # Use the logger run name if the name is not set.
+        if "name" not in self._init_params or self._init_params["name"] is None:
             self._init_params["name"] = logger.run_name
 
+        # Adjust name and group based on `rank_zero_only`.
+        if not self._rank_zero_only:
+            name = self._init_params["name"]
+            group = self._init_params.get("group", None)
+            self._init_params["name"] = f"{name} [RANK_{dist.get_global_rank()}]"
+            self._init_params["group"] = group if group else name
+
         if self._enabled:
             wandb.init(**self._init_params)
 
     def log_file_artifact(self, state: State, log_level: LogLevel, artifact_name: str, file_path: pathlib.Path, *,
                           overwrite: bool):
-        del state, log_level, overwrite  # unused
-
-        # replace all unsupported characters with periods
-        # Only alpha-numeric, periods, hyphens, and underscores are supported by wandb.
-        new_artifact_name = re.sub(r'[^a-zA-Z0-9-_\.]', '.', artifact_name)
-        if new_artifact_name != artifact_name:
-            warnings.warn(("WandB permits only alpha-numeric, periods, hyphens, and underscores in artifact names. "
-                           f"The artifact with name '{artifact_name}' will be stored as '{new_artifact_name}'."))
+        del log_level, overwrite  # unused
 
         if self._enabled and self._log_artifacts:
             import wandb
-            extension = file_path.name.split(".")[-1]
-            artifact = wandb.Artifact(name=new_artifact_name, type=extension)
-            artifact.add_file(os.path.abspath(file_path))
-            wandb.log_artifact(artifact)
+
+            # Some WandB-specific alias extraction
+            timestamp = state.timer.get_timestamp()
+            aliases = ["latest", f"ep{int(timestamp.epoch)}-ba{int(timestamp.batch)}"]
+
+            # replace all unsupported characters with periods
+            # Only alpha-numeric, periods, hyphens, and underscores are supported by wandb.
+            new_artifact_name = re.sub(r'[^a-zA-Z0-9-_\.]', '.', artifact_name)
+            if new_artifact_name != artifact_name:
+                warnings.warn(("WandB permits only alpha-numeric, periods, hyphens, and underscores in artifact names. "
+                               f"The artifact with name '{artifact_name}' will be stored as '{new_artifact_name}'."))
+
+            if self._enabled and self._log_artifacts:
+                import wandb
+                extension = new_artifact_name.split(".")[-1]
+                artifact = wandb.Artifact(name=new_artifact_name, type=extension)
+                artifact.add_file(os.path.abspath(file_path))
+                wandb.log_artifact(artifact, aliases=aliases)
```

Comment:
ooooh, I didn't realize that wandb supports `alias`'s. That's awesome, so we can implement the `symlink_artifact` API for wandb. (Not in this PR; that's for another one).

Reply:
WandB also supports references to external object stores like S3, GCS, which could be a more "enterprise" solution than storing everything on WandB: https://docs.wandb.ai/guides/artifacts/references#s3-gcs-references

This would also simplify multi-gpu and multi-node WandB logging in 99% of cases, because rank0 could build a single artifact that just stores pointers to S3 URIs of each ranks' checkpoints.

Fair warning, this might make your head spin... do `LoggerDestinations` have to know about eachother? Does `ObjectStoreLogger` call `wandb_logger.log_file_artifact(s3_uri)`? Do we implement a 2-stage system like:

```python
# self.storage_logger is a single choice, like ObjectStoreLogger
# self.non_storage_loggers is a list of loggers like WandB, MLFLow, TensorBoard, etc.

device_reference_url = self.storage_logger.log_file_artifact(..., ...)
all_reference_urls = dist.all_gather(device_reference_url)
for logger in self.non_storage_loggers:
  for url in all_reference_urls:
    logger.log_reference(url)
```

So much fun to be had! 🙃